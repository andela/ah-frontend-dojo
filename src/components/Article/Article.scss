@import 'assets/MainStyle.scss';
@import url("https://fonts.googleapis.com/css?family=Lobster&display=swap");
@import "sass-rem";

.canvas {
  background: $white; 
} 

.inputPanel {
  margin-top: 2rem;
  float: left;
  width: 90%;
  margin-left: 5%;
  padding: 0 1.25rem 0 1.25rem;
  height: 40rem;

  &__inputForm {
    width: 70%;
    padding: 0.5rem 0;
    background: #FFFFFF;

    &--input, &--textarea {
      width: 100%;
      border-radius: 0.8rem;
      padding: 0.8rem;
      outline: none;
      margin: 0.4rem 0;
      font-size: 1rem;
      background: $primary-light-2;
      font-family: $primary-font-family;
      border: 0.06rem $primary-light-3 solid;

      &:hover {
        border: 0.05rem $primary solid;
      }

      &:active {
        border: 0.05rem $primary solid;
      }
    }

    #description {
      height: 4rem;
      font-size: 1rem;
    }

    #createArtile {
      padding: 0.2rem 0;
      width: 8rem;
      color: $primary;
      border: 0.06rem $primary solid;
      background: $white;

      &:hover {
        color: $white;
        background: $green;
        border: 0.06rem $green solid;
      }
    }

    &--textarea {
      height: 20rem;
    }
    
    &--button {
      width: 8rem;
      padding: 0.5rem 0.6rem;
      border-radius: 50rem;
      font-size: 1rem;
      outline: none;
      color: $primary;
      border: 0.06rem $primary solid;
      margin: 0.4rem 0;
      cursor: pointer;
      
      &:hover {
        color: $white;
        background: $primary;
      }

      &:active {
        opacity: 0.8;
      }
    }
  }
}
