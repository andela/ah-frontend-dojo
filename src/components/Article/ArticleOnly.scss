@import 'assets/MainStyle.scss';
@import url("https://fonts.googleapis.com/css?family=Lobster&display=swap");
@import "sass-rem";

.writeActive {
  color: $white;
  background: $primary;
}