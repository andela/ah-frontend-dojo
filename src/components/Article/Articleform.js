/* eslint-disable no-unused-vars */
/* eslint-disable react/destructuring-assignment */
import React, { Component } from 'react';
import axios from 'axios';

import './Articles.scss';

class ArticleForm extends Component {
  constructor(props) {
    super(props);
    this.state = {};
  }

  render() {
    return (
      <div>
        <h3>Add an article</h3>

      </div>
    );
  }
}

export default ArticleForm;
