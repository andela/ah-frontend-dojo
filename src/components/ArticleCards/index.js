// react libraries
import React, { Component } from 'react';

// styles
import './ArticleCardLarge.scss';

// components
import Tagslist from 'components/TagsList';
import Comments from 'components/Comments';
import PropTypes from 'prop-types';

// helper function
import { formatDate, isAuthenticated } from 'utils';

export const checkOwner = (author, article_id, deleteArticle) => {
  if (isAuthenticated().username === author) {
    return (
      <div className="protectedButtons" align="right">
        <a href={`/article_edit/${article_id}`} className="blue">Edit</a>
        <form onSubmit={deleteArticle}>
          <button type="submit" className="red">Delete</button>
        </form>
      </div>
    );
  }
  return ('');
};

export default class ArticleCardLarge extends Component {
  constructor(props) {
    super(props);
    this.props = {
      article: {
        id: 0,
        title: '',
        body: '',
        description: '',
        author: '',
        updatedAt: '',
        time_to_read: '',
      }
    };
  }

  onDeleteArticle = (e) => {
    e.preventDefault();
  };

  render() {
    const {
      article: {
        id, title, description, body, updatedAt, author, time_to_read
      }
    } = this.props;
    return (
      <div className="mainCanvas">
        <div className="mainCanvas__cardCanvas">
          <div className="mainCanvas__cardCanvas__titlePanel">
            <p>{ title }</p>
          </div>
          <div className="mainCanvas__cardCanvas__titlePanel__imagePanel">
            <img
              src="https://res.cloudinary.com/andelaprojects/image/upload/v1561099197/banner_hrpjdt.jpg"
              alt="article img"
            />
          </div>
          <div className="mainCanvas__cardCanvas__titlePanel__flexedRow">
            <div className="mainCanvas__cardCanvas__titlePanel__flexedRow__columnPanel">
              <div className="mainCanvas__cardCanvas__titlePanel__flexedRow__columnPanel__aurthorPanel">
                <img
                  className="mainCanvas__cardCanvas__titlePanel__flexedRow__columnPanel__aurthorPanel--authorAvartar"
                  src="https://res.cloudinary.com/andelaprojects/image/upload/v1561099197/banner_hrpjdt.jpg"
                  alt="article img"
                />
                <div
                  className="mainCanvas__cardCanvas__titlePanel__flexedRow__columnPanel__aurthorPanel--authorName"
                >
                  <p align="left">
                    <b>{ author }</b>
                    <br />
                    <button type="submit" className="buttonSmall buttonLarge">
                      View profile
                    </button>
                  </p>
                </div>
              </div>
            </div>
            <div className="mainCanvas__cardCanvas__titlePanel__flexedRow__columnPanel">
              <div className="mainCanvas__cardCanvas__titlePanel__flexedRow__columnPanel__articleDetails_1 align-inText-right lightedParagraph">
                <p>
                  <b>Date: </b>
                  { formatDate(updatedAt) }
                </p>
                <p>
                  <b>Read time: </b>
                  { time_to_read }
                </p>
              </div>
            </div>
          </div>
          <div className="bodyPanel">
            <p className="description">{ description }</p>
            <p>{ body }</p>
          </div>
          <div className="additionalDetails inlinePosition">
            <p>
              <b>Average Rating: </b>
              3.9 (2000)
            </p>
            <p>
              <b>Views: </b>
              (0000)
            </p>
            <p>
              <b>Likes: </b>
              (0000)
            </p>
            <p>
              <b>Dislikes: </b>
              (0000)
            </p>
          </div>
          <Tagslist />
          {checkOwner(author, id, this.del)}
          <div className="splitLine" />
          <div className="commentsPanel">
            <Comments />
          </div>
        </div>
      </div>
    );
  }
}

ArticleCardLarge.propTypes = {
  article: PropTypes.shape({
    id: PropTypes.number,
    title: PropTypes.string,
    body: PropTypes.string,
    description: PropTypes.string,
    author: PropTypes.string,
    updatedAt: PropTypes.string,
    time_to_read: PropTypes.number,
  }),
};

ArticleCardLarge.defaultProps = {
  article: {
    id: 0,
    title: '',
    body: '',
    description: '',
    author: '',
    updatedAt: '',
    time_to_read: '',
  }
};
