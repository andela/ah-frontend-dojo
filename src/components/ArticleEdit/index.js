// react libraries
import React, { Component } from 'react';

// third-party libraries
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

// styles
import 'components/Loader/Loader.scss';
import 'components/Footer/Footer.scss';
import './ArticleEdit.scss';

// components
import Navbar from 'components/NavBar';
import Footer from 'components/Footer';
import { getOneArticle, editArticle } from 'store/actions/articleActions';
import ProgressLoader from 'components/Loader/ProgressLoader';


export class ArticleEdit extends Component {
  constructor(props) {
    super(props);
    this.state = {
      title: '',
      description: '',
      body: '',
      tagList: [],
    };
  }

  componentDidMount() {
    const { getArticle, match: { params: { id } }, article } = this.props;
    getArticle(id);

    this.setState(prevState => ({ ...prevState, article }));
  }


  handleChange = (e) => {
    this.setState({ [e.target.name]: e.target.value });
  }

  handleSubmit = (e) => {
    e.preventDefault();
    const {
      title,
      description,
      body
    } = this.state;
    const articleData = {
      title,
      description,
      body,
      tagList: [],
    };
    const { updateArticle, article } = this.props;

    if (articleData.title === '') {
      articleData.title = article.title;
    }
    if (articleData.description === '') {
      articleData.description = article.description;
    }
    if (articleData.body === '') {
      articleData.body = article.body;
    }

    const articleObject = {
      id: article.id,
      article: articleData,
    };
    updateArticle(articleObject);
  }

  cancelArticleEdit = () => {
    const { article } = this.props;
    window.location = `/article/${article.id}`;
  }

  render() {
    const { article, editProgress } = this.props;
    return (
      <div className="canvas">
        <div className="landing">
          <Navbar />
          <div className="breadcrumbs" align="left">
            <ul>
              <li>
                <img
                  src="https://res.cloudinary.com/andelaprojects/image/upload/v1561099269/breadcrumbs_qot92i.png"
                  alt="breadcrumbs"
                />
                /Edit Article/
              </li>
            </ul>
          </div>
          {editProgress && <ProgressLoader />}
          <div className="inputPanel" align="center">
            <div className="inputForm">
              <form onSubmit={this.handleSubmit}>
                <input type="text" name="title" defaultValue={article.title} onChange={this.handleChange} />
                <textarea
                  type="text"
                  name="description"
                  id="description"
                  onChange={this.handleChange}
                  placeholder={article.description}
                />
                <textarea
                  type="text"
                  name="body"
                  value={article.body}
                />
                <div className="editButtons">
                  <button
                    type="submit"
                    className=""
                    id="edit"
                    name="updateArticle"
                  >
                      Save Changes
                  </button>
                  <button type="button" className="" id="cancel" onClick={this.cancelArticleEdit} name="cancel">Cancel</button>
                </div>
              </form>
            </div>
          </div>
          <Footer />
        </div>
      </div>
    );
  }
}

ArticleEdit.propTypes = {
  getArticle: PropTypes.func.isRequired,
  updateArticle: PropTypes.func.isRequired,
  editProgress: PropTypes.bool,
  match: PropTypes.shape(
    {
      params: PropTypes.shape(
        {
          id: PropTypes.isRequired,
        }
      )
    }
  ).isRequired,
  article: PropTypes.shape({
    id: PropTypes.number,
    title: PropTypes.string,
    description: PropTypes.string,
    body: PropTypes.string,
  })
};

ArticleEdit.defaultProps = {
  editProgress: false,
  article: {
    id: 61,
    title: '',
    description: '',
    body: '',
  }
};

export const mapStateToProps = (state) => {
  const { article, editProgress } = state.articleReducer;
  const articleId = state.articleReducer.article.id;
  return { article, articleId, editProgress };
};

export const mapDispatchToProps = dispatch => ({
  getArticle: (articleData) => { dispatch(getOneArticle(articleData)); },
  updateArticle: (articleData) => { dispatch(editArticle(articleData)); }
});

export default connect(mapStateToProps, mapDispatchToProps)(ArticleEdit);
