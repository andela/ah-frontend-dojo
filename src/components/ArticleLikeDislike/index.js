// react libraries
import React, { Component } from 'react';

// third party libraries
import IosThumbsUpOutline from 'react-ionicons/lib/IosThumbsUpOutline';
import IosThumbsDownOutline from 'react-ionicons/lib/IosThumbsDownOutline';

// styles
import './ArticleLikeDislike.scss';

export default class ArticleLikeDislike extends Component {
  render() {
    return (
      <div className="likeDislikePanel">
        <div className="like">
          <IosThumbsUpOutline color="#2D6686" fontSize="1.5rem" cursor="pointer" />
          <b>Likes: </b>
        </div>
        <div className="dislike">
          <IosThumbsDownOutline color="#2D6686" />
          <b>Dislikes: </b>
        </div>
      </div>
    );
  }
}
