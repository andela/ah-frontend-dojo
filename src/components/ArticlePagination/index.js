// react
import React from 'react';

// Proptypes for validation
import PropTypes from 'prop-types';

// stylesheets
import './ArticlePagination.scss';

// Button component
import Button from 'components/Button';

const ArticlePagination = ({
  previous, next, paginate, count, page
}) => {
  const pages = Math.ceil(count / 10);
  return (

    <div className="article-pagination">
      <Button
        btnName=""
        btnClass="fa fa-backward"
        btnEvent={(e) => {
          e.preventDefault();
          paginate(previous, -1);
        }}
        disabled={!previous}
      />
      <span>
        {page || 1}
        {' '}
        {'/'}
        {' '}
        {pages}
        {' '}
        {'pages '}
      </span>
      <Button
        btnName=""
        btnClass="fa fa-forward"
        btnEvent={(e) => {
          e.preventDefault();
          paginate(next, 1);
        }}
        disabled={!next}
      />
    </div>
  );
};

ArticlePagination.defaultProps = {
  next: null,
  previous: null,
};

ArticlePagination.propTypes = {
  next: PropTypes.string,
  page: PropTypes.number.isRequired,
  count: PropTypes.number.isRequired,
  previous: PropTypes.string,
  paginate: PropTypes.func.isRequired,
};


export default ArticlePagination;
