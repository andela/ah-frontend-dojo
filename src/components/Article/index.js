// react libraries
import React, { Component } from 'react';

// third-party libraries
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

// styles
import 'components/Loader/Loader.scss';
import 'components/Footer/Footer.scss';
import './Article.scss';

// components
import Navbar from 'components/NavBar';
import Footer from 'components/Footer';
import { newArticle } from 'store/actions/articleActions';


export class Article extends Component {
  constructor(props) {
    super(props);
    this.state = {
      title: '',
      description: '',
      body: '',
    };
  }

  handleChange = (e) => {
    this.setState({ [e.target.name]: e.target.value });
  }

  handleSubmit = (e) => {
    e.preventDefault();
    const { title, description, body } = this.state;
    const articleData = {
      article: {
        title,
        description,
        body,
        tagList: [],
      },
    };

    const { createArticle } = this.props;

    createArticle(articleData);
  }

  render() {
    const { title, description, body } = this.state;
    return (
      <div className="canvas">
        <div className="landing">
          <Navbar />
          <div className="breadcrumbs" align="left">
            <ul>
              <li>
                <img src="https://res.cloudinary.com/andelaprojects/image/upload/v1561099269/breadcrumbs_qot92i.png" alt="breadcrumbs" />
                /Create Article/
              </li>
            </ul>
          </div>
          <div className="inputPanel" align="center">
            <div className="inputPanel__inputForm">
              <form onSubmit={this.handleSubmit}>
                <input
                  className="inputPanel__inputForm--input"
                  type="text"
                  name="title"
                  placeholder="Title"
                  value={title}
                  onChange={this.handleChange}
                  required
                />
                <input
                  className="inputPanel__inputForm--input"
                  type="text"
                  id="description"
                  name="description"
                  placeholder="Description"
                  value={description}
                  onChange={this.handleChange}
                  required
                />
                <textarea
                  className="inputPanel__inputForm--textarea"
                  type="text"
                  name="body"
                  placeholder="Body"
                  value={body}
                  onChange={this.handleChange}
                  required
                />
                <button
                  className="inputPanel__inputForm--button"
                  type="submit"
                  name="createArtile"
                >
                  Create Article
                </button>
              </form>
            </div>
          </div>
          <Footer />
        </div>
      </div>
    );
  }
}

Article.propTypes = {
  createArticle: PropTypes.func.isRequired,
};

export const mapStateToProps = ({ articleReducer: article }) => (article);

export const mapDispatchToProps = dispatch => ({
  createArticle(articleData) { dispatch(newArticle(articleData)); }
});

export default connect(mapStateToProps, mapDispatchToProps)(Article);
