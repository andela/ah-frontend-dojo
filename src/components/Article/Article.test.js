import React from 'react';
import { shallow } from 'enzyme';
import { Article, mapStateToProps, mapDispatchToProps } from 'components/Article/';

// test to check if the Article component matches the snapshot
describe('Article Component', () => {
  const props = {
    createArticle: jest.fn(),
  };
  const component = shallow(<Article {...props} />);
  component.setState({
    title: '',
    description: '',
    body: '',
    tagList: [],
  });

  it('matches the snapshot', () => {
    expect(component).toMatchSnapshot();
  });

  it('should update title state on input change', () => {
    component.find('input[name="title"]').simulate(
      'change',
      {
        target: { name: 'title', value: 'article title' }
      }
    );

    expect(component.state('title')).toEqual('article title');
  });

  it('should update description state on input change', () => {
    component.find('input[name="description"]').simulate(
      'change',
      {
        target: { name: 'description', value: 'article description' }
      }
    );

    expect(component.state('description')).toEqual('article description');
  });

  it('should update body state on input change', () => {
    component.find('textarea[name="body"]').simulate(
      'change',
      {
        target: { name: 'body', value: 'article body' }
      }
    );

    expect(component.state('body')).toEqual('article body');
  });

  it('should handle input change event', () => {
    const event = {
      target: {
        name: 'title',
        value: 'new article title',
      },
    };
    component.instance().handleChange(event);
    expect(component.instance().state.title).toBe('new article title');
  });

  it('should handle submit event', () => {
    const article = {
      title: 'title text',
      description: 'description text',
      body: 'body text',
    };
    component.setState(article);
    const event = {
      target: {
        type: 'submit',
        name: 'createArtile',
      },
      preventDefault: jest.fn(),
    };
    component.instance().handleSubmit(event);
    expect(props.createArticle).toBeCalled();
    article.tagList = [];
    expect(props.createArticle).toBeCalledWith({ article });
  });

  it('should map state to props', () => {
    const article = {
      title: 'title text',
      description: 'description text',
      body: 'body text',
      target: [],
    };

    const mockedState = {
      articleReducer: { article }
    };

    const state = mapStateToProps((mockedState));
    expect(state.article.title).toBe('title text');
  });

  it('should map dispatch to props', () => {
    const mockDispatch = jest.fn();

    mapDispatchToProps((mockDispatch)).createArticle();
    expect(mockDispatch).toHaveBeenCalled();
  });
});
