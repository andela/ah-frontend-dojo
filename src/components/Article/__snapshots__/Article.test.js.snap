// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Article Component matches the snapshot 1`] = `
<div
  className="canvas"
>
  <div
    className="landing"
  >
    <Connect(Navbar) />
    <div
      align="left"
      className="breadcrumbs"
    >
      <ul>
        <li>
          <img
            alt="breadcrumbs"
            src="https://res.cloudinary.com/andelaprojects/image/upload/v1561099269/breadcrumbs_qot92i.png"
          />
          /Create Article/
        </li>
      </ul>
    </div>
    <div
      align="center"
      className="inputPanel"
    >
      <div
        className="inputPanel__inputForm"
      >
        <form
          onSubmit={[Function]}
        >
          <input
            className="inputPanel__inputForm--input"
            name="title"
            onChange={[Function]}
            placeholder="Title"
            required={true}
            type="text"
            value=""
          />
          <input
            className="inputPanel__inputForm--input"
            id="description"
            name="description"
            onChange={[Function]}
            placeholder="Description"
            required={true}
            type="text"
            value=""
          />
          <textarea
            className="inputPanel__inputForm--textarea"
            name="body"
            onChange={[Function]}
            placeholder="Body"
            required={true}
            type="text"
            value=""
          />
          <button
            className="inputPanel__inputForm--button"
            name="createArtile"
            type="submit"
          >
            Create Article
          </button>
        </form>
      </div>
    </div>
    <Footer />
  </div>
</div>
`;
